app/models/notification.rb

```diff
@@ -54,6 +56,15 @@ def self.notification_text(name, message_params)
 
   private
 
+  def complete_bindings
+    sprintf(notification_type.message, to_h[:bindings])
+  rescue ArgumentError, KeyError => e
+    # 1. Deprecate now
+    # 2. Fail validation going forward via errors.add(error_args)
+    error_args = [:options, "text bindings for notification_type: '#{notification_type.name}' failed with error: '#{e.message}' with options: '#{options.inspect}' and message #{notification_type.message.inspect}. Next release will not allow a notification without complete bindings."]
+    ActiveSupport::Deprecation.warn("#{error_args.join(' ')}",  caller_locations[1..-1].reject {|location| location.label.include?("emit_for_event")})
```

Comment:
Note, `emit_for_event` above is the primary caller where notifications get created indirectly if the event has a matching `notification_type` when you raise an event so we have to skip it as a possible `caller` since that's not where you need to fix the code to provide a complete event, including the `full_data` containing the values to substitute.  It's complicated, but that's how it works currently.

Reply:
I would vote to put some sort of comment in the code directly about it then so the reasoning isn't lost to time.